File: src/api/daemon/member.ts
/**
 * Daemon member API module — typed HTTP client for per-member sync preferences.
 *
 * Phase 4b PR-3：取代 `./device.ts` 里读写 `PairedDevice.sync_settings` 的旧路径,
 * 对应后端 `MemberSyncPreferencesDto` (双向 send/receive + 双套 content types)。
 *
 * # Endpoints / 端点
 * - `GET /member/:device_id/sync-preferences`
 * - `PATCH /member/:device_id/sync-preferences`
 *
 * 本阶段前端 UX 只展示 send 列；receive 字段不露出,由后端默认值保持
 * (`MemberSyncPreferences::default()`: `send/receive = true`,两套 `ContentTypes` 全开)。
 * 未来若要暴露接收开关,直接扩展本模块与 `DeviceSettingsSheet` 即可。
 */

import {
  getMemberSyncPreferences as getMemberSyncPreferencesSdk,
  updateMemberSyncPreferences as updateMemberSyncPreferencesSdk,
} from '@/api/generated/sdk.gen'
import type { MemberSyncPreferencesPatchDto } from '@/api/generated/types.gen'
import { daemonClient } from './client'

// ── Value objects ───────────────────────────────────────────────

/** Content type toggles. Matches `ContentTypesDto` on the Rust side. */
export interface ContentTypes {
  text: boolean
  image: boolean
  link: boolean
  file: boolean
  codeSnippet: boolean
  richText: boolean
}

/** Partial content type toggles for PATCH. */
export interface ContentTypesPatch {
  text?: boolean
  image?: boolean
  link?: boolean
  file?: boolean
  codeSnippet?: boolean
  richText?: boolean
}

/**
 * Sync preferences recorded for a space member.
 * Matches `MemberSyncPreferencesDto` on the Rust side.
 */
export interface MemberSyncPreferences {
  sendEnabled: boolean
  receiveEnabled: boolean
  sendContentTypes: ContentTypes
  receiveContentTypes: ContentTypes
}

/**
 * Partial member sync preferences for PATCH.
 * Any omitted top-level field keeps its current value server-side.
 */
export interface MemberSyncPreferencesPatch {
  sendEnabled?: boolean
  receiveEnabled?: boolean
  sendContentTypes?: ContentTypesPatch
  receiveContentTypes?: ContentTypesPatch
}

// ── Public API ──────────────────────────────────────────────────

export async function getMemberSyncPreferences(deviceId: string): Promise<MemberSyncPreferences> {
  // Route through the generated SDK; `callEnveloped` unwraps the SDK's `{ data }`
  // envelope down to the preferences payload. The generated
  // `MemberSyncPreferencesDto` is structurally equivalent to the hand-written
  // `MemberSyncPreferences` (camelCase wire fields), bridged here to keep the
  // public return type stable for downstream consumers.
  const data = await daemonClient.callEnveloped(() =>
    getMemberSyncPreferencesSdk({ path: { device_id: deviceId }, throwOnError: true })
  )
  return data as unknown as MemberSyncPreferences
}

export async function updateMemberSyncPreferences(
  deviceId: string,
  patch: MemberSyncPreferencesPatch
): Promise<MemberSyncPreferences> {
  // PATCH returns only `{ data: { success } }` (ApiEnvelope<MemberSyncResultDto>);
  // callSdk throws on non-2xx, so reaching here means success. Re-fetch to return
  // the authoritative merged preferences (preserves the Promise<MemberSyncPreferences>
  // contract that devicesSlice stores into state — the PATCH body no longer echoes it).
  await daemonClient.callSdk(() =>
    updateMemberSyncPreferencesSdk({
      path: { device_id: deviceId },
      // `MemberSyncPreferencesPatch` is structurally equivalent to the generated
      // `MemberSyncPreferencesPatchDto`; bridge for the SDK body param.
      body: patch as unknown as MemberSyncPreferencesPatchDto,
      throwOnError: true,
    })
  )
  return getMemberSyncPreferences(deviceId)
}

// ── Default-value helpers (used by "restore defaults" button) ──

/**
 * Mirror of `MemberSyncPreferences::default()` on the Rust side (all toggles on).
 * Used by Restore Defaults to push an explicit reset without relying on a
 * server-side "null clears overrides" semantic (which does not exist for
 * `space_member`; every member always has a preferences record).
 */
export const DEFAULT_SEND_CONTENT_TYPES: ContentTypes = {
  text: true,
  image: true,
  link: true,
  file: true,
  codeSnippet: true,
  richText: true,
}

File: src-tauri/crates/uc-daemon-contract/src/api/dto/member.rs
//! DTOs for per-member sync preferences (phase 4b PR-2).
//!
//! 语义：映射 `SpaceMember.sync_preferences`（双向 `send_enabled` /
//! `receive_enabled` + 双套 `content_types`）。复用 `dto::settings` 下的
//! `ContentTypesDto` / `ContentTypesPatchDto`，两套字段形状一致。

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use uc_core::membership::MemberSyncPreferences;

use super::settings::{ContentTypesDto, ContentTypesPatchDto};

/// Sync preferences recorded for a space member.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemberSyncPreferencesDto {
    pub send_enabled: bool,
    pub receive_enabled: bool,
    pub send_content_types: ContentTypesDto,
    pub receive_content_types: ContentTypesDto,
}

/// Partial sync preferences for PATCH /member/:device_id/sync-preferences.
///
/// 服务器侧 `get → merge → save` 后持久化；未提供的字段保留当前值。
/// 重置到默认值的调用方应显式传入所有字段的默认值（`MemberSyncPreferences::default()`）。
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemberSyncPreferencesPatchDto {
    pub send_enabled: Option<bool>,
    pub receive_enabled: Option<bool>,
    pub send_content_types: Option<ContentTypesPatchDto>,
    pub receive_content_types: Option<ContentTypesPatchDto>,
}

/// Folded payload for `PATCH /member/:device_id/sync-preferences` (ADR-008 §0.1).
///
/// The current handler returns `success` as a top-level sibling of the
/// `{data,ts}` envelope. This DTO folds it INTO the payload so the endpoint can
/// return `ApiEnvelope<MemberSyncResultDto>` with no bespoke wrapper. P1 only
/// defines the type; the handler is rewired in P2.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemberSyncResultDto {
    pub success: bool,
}

impl From<MemberSyncPreferences> for MemberSyncPreferencesDto {
    fn from(value: MemberSyncPreferences) -> Self {
        Self {
            send_enabled: value.send_enabled,
            receive_enabled: value.receive_enabled,
            send_content_types: value.send_content_types.into(),
            receive_content_types: value.receive_content_types.into(),
        }
    }
}
